Argument Relevance for Synthesis Conjectures

ajreynol opened the pull request:
EDIT : merged pull request #1323, to take care of part of this pull request.

This introduces a new algorithm for detecting when argument positions of functions-to-synthesize are irrelevant.  It is a working version of a solution for #1210.

The algorithm is implemented in the CegProcessConj class, which is invoked when a synthesis conjecture is registered.  Based on an analysis, the grammar construction algorithm in CegGrammarCons can restrict the grammar for the function-to-synthesize.

TODO:
- Symmetry breaking for user-provided grammars,
- Generalize the cases covered by the algorithm.

ajreynol commented:
I've taken some of this PR and added it to #1323 to expedite getting the changes into master that are needed for Haniel to get started working, so let's merge #1323 before this one.

mpreiner commented on src/theory/quantifiers/sygus_process_conj.cpp:
```diff
@@ -27,11 +27,546 @@ namespace CVC4 {
+        for (std::unordered_set<Node, NodeHashFunction>::iterator itfv =
+                 itf->second.begin();
+             itfv != itf->second.end();
+             ++itfv)
```
This is again a case where a for each loop would be more compact and easier to read.

mpreiner commented on src/theory/quantifiers/sygus_process_conj.cpp:
```diff
@@ -27,11 +27,546 @@ namespace CVC4 {
+  // process the conjunctions
+  for (std::map<Node, CegConjectureProcessFun>::iterator it = d_sf_info.begin();
+       it != d_sf_info.end();
+       ++it)
```
Ditto (for each).

mpreiner commented on src/theory/quantifiers/sygus_process_conj.cpp:
```diff
@@ -47,51 +582,223 @@ void CegConjectureProcess::initialize(Node n, std::vector<Node>& candidates)
-    base = n[0][1];
+    return its->second.isArgRelevant(i);
   }
   else
```
Remove `else`?

mpreiner commented on src/theory/quantifiers/sygus_process_conj.cpp:
```diff
@@ -47,51 +582,223 @@ void CegConjectureProcess::initialize(Node n, std::vector<Node>& candidates)
+    its->second.getIrrelevantArgs(args);
+    return true;
   }
   else
```
Ditto.

mpreiner commented on src/theory/quantifiers/sygus_process_conj.h:
```diff
@@ -15,26 +15,234 @@
+ *
+ * The classes below compute whether certain arguments of
+ * the function-to-synthesize f are irrelevant.
+ * Assume that f is binary function, where possible solutions
```
Typo: ... is **a** binary ...

mpreiner commented on src/theory/quantifiers/sygus_process_conj.h:
```diff
@@ -15,26 +15,234 @@
+                  std::unordered_map<unsigned, Node>& n_arg_map);
+  /** infer definition
+   *
+   * In the following, we say a term is a "emplate
```
Typo: ... "**t**emplate definition" ...

mpreiner commented on src/theory/quantifiers/sygus_process_conj.h:
```diff
@@ -15,26 +15,234 @@
+  /** infer definition
+   *
+   * In the following, we say a term is a "emplate
+   * definition" if its free variables
```
Remove newline.

mpreiner commented on src/theory/quantifiers/sygus_process_conj.h:
```diff
@@ -15,26 +15,234 @@
+   * If def is non-null,
+   * this function assigns def as a template definition
+   * for the (currently irrelevant) arguments that
+   * occur in a term of the form f(
```
There seems to be something missing: ... of the form f( ?

ajreynol replied:
Yes, I'll fix.

mpreiner commented on src/theory/quantifiers/sygus_process_conj.h:
```diff
@@ -15,26 +15,234 @@
+   * If def is null, then an argument is marked
+   * as relevant.
+   *
+   * Returns an value rid such that:
```
Typo: ... **a** value ...

ajreynol commented:
I've addressed the review in my latest commit.  I'd still like to hold off on for -> for each unless there is a consensus decision in the style guidelines.

ajreynol commented on src/theory/quantifiers/ce_guided_pbe.h:
```diff
@@ -192,8 +190,7 @@ class CegConjecturePbe {
   /** parent conjecture
-  * This is the data structure that contains global information about the
-  * conjecture.
+  * This contains global information about the synthesis conjecture.
```
This comment was out of place, I've replaced it.
